### ISSUE TITLE
unable to upload music into catalog

### ISSUE THREAD

**Issue by NatureMuncher** [ampache/ampache#2053]
# Issue Template

## Environment

* Ampache version: Ampache 4.0.0-develop
* Web server + version: Apache/2.4.6 
* Server operating system: CentOS Linux release 7.7.1908
* Client operating system: Fedora release 30 (Thirty)

* Client type:
  * Web: Chrome Version 77.0.3865.75 (Official Build) beta (64-bit) 

## Settings

See https://pastebin.com/aCNGYefN

## Description

All is working well except the ability to upload music. It acts like it uploads but nothing shows under Uploads and the music cannot be located in the catalog.
I did follow the steps to set `Uploads catalog destination` to the existing catalog.
Also performed the Catalog "Add" function to re-read the contents as suggested but each time it seems to get distracted by finding cover art for various albums. My new music never appears.

I did notice that a folder exists under the catalog with my name and the datestamp when I initially tried to upload, as would be expected since I left `Upload: create a subdirectory per user (recommended)` at the default setting of Enabled.

* SELinux is in permissive mode.
* Permissions on the catalog root have been set to `u+rwX,go+rX` per https://github.com/ampache/ampache/wiki/Catalog#permissions
* Ownership on the catalog root has been set to `apache:apache`

## Logs
From ampache logs:
```
2019-09-20 20:39:37 [ampache] (session.class) -> Existing session found 
2019-09-20 20:39:37 [mike] (session.class) -> qjb0gsqtbgnplq14bisgsqhcn7 has been extended to Fri, 20 Sep 2019 21:39:37 +0000 extension length 3600 
2019-09-20 20:39:37 [mike] (ajax.server) -> Called for page: {index} 
2019-09-20 20:39:40 [ampache] (session.class) -> Existing session found 
2019-09-20 20:39:40 [mike] (session.class) -> qjb0gsqtbgnplq14bisgsqhcn7 has been extended to Fri, 20 Sep 2019 21:39:40 +0000 extension length 3600 
2019-09-20 20:39:40 [mike] (ajax.server) -> Called for page: {index} 
2019-09-20 20:39:44 [ampache] (session.class) -> Existing session found 
2019-09-20 20:39:44 [mike] (session.class) -> qjb0gsqtbgnplq14bisgsqhcn7 has been extended to Fri, 20 Sep 2019 21:39:44 +0000 extension length 3600 
2019-09-20 20:39:44 [mike] (session.class) -> Session created: 3bc67d9e907767d24f32b0d543d27a66 
2019-09-20 20:39:44 [ampache] (session.class) -> Existing session found 
2019-09-20 20:39:44 [mike] (session.class) -> qjb0gsqtbgnplq14bisgsqhcn7 has been extended to Fri, 20 Sep 2019 21:39:44 +0000 extension length 3600 
2019-09-20 20:40:03 [ampache] (session.class) -> Existing session found 
2019-09-20 20:40:03 [mike] (session.class) -> qjb0gsqtbgnplq14bisgsqhcn7 has been extended to Fri, 20 Sep 2019 21:40:03 +0000 extension length 3600 
2019-09-20 20:40:03 [mike] (session.class) -> Session created: 0e37fedff9265d5769ce442f4f05ecf4 
```
from apache access_log:
```
192.168.192.168 - - [20/Sep/2019:15:39:40 -0500] "POST /ampache/server/ajax.server.php?page=index&action=sidebar&button=home HTTP/1.1" 200 11244 "http://ampache-server.domain.com/ampache/index.php" "Mozilla/5.0 (X11; Fedora; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.75 Safari/537.36"
192.168.192.168 - - [20/Sep/2019:15:39:44 -0500] "GET /ampache/upload.php HTTP/1.1" 200 265914 "http://ampache-server.domain.com/ampache/index.php" "Mozilla/5.0 (X11; Fedora; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.75 Safari/537.36"
192.168.192.168 - - [20/Sep/2019:15:39:44 -0500] "GET /ampache/server/fs.ajax.php?operation=get_node&id=%23 HTTP/1.1" 200 97 "http://ampache-server.domain.com/ampache/index.php" "Mozilla/5.0 (X11; Fedora; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.75 Safari/537.36"
192.168.192.168 - - [20/Sep/2019:15:40:03 -0500] "POST /ampache/upload.php HTTP/1.1" 200 265914 "http://ampache-server.domain.com/ampache/index.php" "Mozilla/5.0 (X11; Fedora; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.75 Safari/537.36"
192.168.192.168 - - [20/Sep/2019:15:40:37 -0500] "POST /ampache/server/ajax.server.php?page=index&action=reloadnp HTTP/1.1" 200 35093 "http://ampache-server.domain.com/ampache/index.php" "Mozilla/5.0 (X11; Fedora; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.75 Safari/537.36"
```
apache error_log is empty

**Comment by manuelglez86** [comment]
Did you configure the php.ini? You need to set the upload_max_filesize and the post_max_size, Did you set Allowoverride All on the httpd.conf?

**Comment by NatureMuncher** [comment]
Yikes! No I think I missed those in the guide. I'll give it a look. Thanks
for this!

**Comment by NatureMuncher** [comment]
Thanks for the suggestions! I confirmed the php.ini settings were already set to 20M and 200M respectively. Also set timezone since I missed that.
I added the Allowoverride for the ampache path and restarted apache:
```
<Directory /var/www/html/ampache>
    AllowOverride All
    Require all granted
</Directory>
```
However even before that I have a bigger issue. Users can still log in but they're all redirected to an unprivileged account. Here's the logfile during a login attempt:
https://pastebin.com/aVfCZGkM

**Comment by manuelglez86** [comment]
Allowoveride All goes under the default site, 

```
<Directory /var/www/html/>
     AllowOverride All
     Require all granted
 </Directory>
```
maybe setting it to the ampache folcder is what is causing the error

**Comment by NatureMuncher** [comment]
Oh I added it as a separate block below the default site at first but just tried your way and it appears to be doing the same.
I think I'll open a separate issue to fix the logins then come back to this, since I can't upload anything as a guest user anyway...

**Comment by NatureMuncher** [comment]
I got my authentication issue [worked out](https://github.com/ampache/ampache/issues/2056#issuecomment-542795891) but still cannot upload music. Attached is a pic of the upload screen. I have songs and cover art selected, but there is no "save" button. If I click the green checkmarks the songs disappear from the list. What am I missing?
![upload](https://user-images.githubusercontent.com/1648745/67020802-53b2e500-f0c4-11e9-909d-1d33e1e826eb.png)

**Comment by manuelglez86** [comment]
The green checkmarks usually means that the songs uploaded to the server, did you try selectting or creating an album and an artist? did you set the upload catalog destination? Does the folder for uploads have the appropiatte permissions for apache?

**Comment by NatureMuncher** [comment]
So the checkmarks indicate that the file was uploaded successfully? That wasn't clear to me.
Then all I need to do is to navigate to the admin panel, open the catalog and click "ADD", right?
Tried all that but still nada. Tried searching for the songs, checked my uploads page but nothing there either.

**Comment by wagnered** [comment]
If I remember correctly, uploaded files should be listed on the home page under information->uploads.  If you directed ampache to create a subfolder, the file should physically reside in a subfolder with your username.  I guess it was done that way to keep them private.  It is my opinion that the user should have the choice to make it public or private when uploading it.



-------- Original message --------
Subject: Re: [ampache/ampache] unable to upload music into catalog (#2053)

**Comment by lachlan-00** [comment]
I will need to test all this as I do uploads outside Ampache using rsync.

The entire process could probably be a bit more usable

**Comment by wagnered** [comment]
Oops, sorry @NatureMuncher, just noticed that you did check the Information->Uploads page.



-------- Original message --------
Subject: Re: [ampache/ampache] unable to upload music into catalog (#2053)

**Comment by NatureMuncher** [comment]
Thank you for looking into it. I'm glad to provide testing and logs to get this worked out. Uploading is the last piece before I'm ready to share Ampache with my coworkers.

**Comment by manuelglez86** [comment]
When you upload a song the actual file is saved in the same folder of the catalog you choose as upload catalog, can you chek if the uploads files are present (if active ampache creates a folder with the user name), if the files are not present then the problem could be a configuration or permissions issue rather than an ampache bug

**Comment by NatureMuncher** [comment]
Yes a folder is created with my name but it remains empty despite several attempts at uploading. Permissions look good to me:
```
[root@hostname ~]# ls -lhat /home/sounds/mp3/music/mike/
total 52K
drwxrwxr-x. 548 apache apache 32K Sep 19 11:34 ..
drwxr-xr-x.   2 apache apache   6 Sep 19 11:34 .
```
And the parent directory shows the last added music was from April, before we had Ampache working:
```
[root@hostname ~]# ls -lhat /home/sounds/mp3/music/|head
total 184M
drwxrwxr-x. 548 apache apache  32K Sep 19 11:34 .
drwxr-xr-x.   2 apache apache    6 Sep 19 11:34 mike
drwxrwxr-x.   4 sounds apache   30 Apr  3  2019 ..
-rwxr-xr-x.   1 apache apache 4.8M Apr  3  2019 Europe-The_Final_Countdown.mp3.mpga
-rwxr-xr-x.   1 apache apache  728 Apr  3  2019 golden_smog-another_fine_day.m3u
-rwxr-xr-x.   1 apache apache  768 Apr  3  2019 gin_blossoms-congratulations_im_sorry.m3u
```

**Comment by manuelglez86** [comment]
theres something on ampche logs when you upload a file?

**Comment by manuelglez86** [comment]
did you try to upload to a folder outsite of the home directory?, I recently have an issue where I try to upload albumart and set the destinetion to a folder on the home directory and it never upload, I change to a folder on the hdd (/albumarts in my case) and then I could upload the images.

Also check your firewall (if you have one)

**Comment by NatureMuncher** [comment]
Thanks for the tip. Tried toggling off the "Create a subdirectory per user" option and it then provided a drop-down for me to pick a folder. Did so, selected the items for upload and got green checkmarks but nothing new exists within that folder:
```
[root@hostname ~]# ls -Artls /home/geoff/mp3/music/311 | tail -n 1                                                                                                                                                                  
   4 drwxrwxr-x. 2 apache apache    4096 Oct  3  2017 Grassroots
```
Believe the upload is performed via the browser so ports 80/443 are open.

**Comment by lachlan-00** [comment]
I haven't started on this yet but will be adding debugging/checks in the process.

@wagnerd is probably right in the the process is sending a false/null and not telling you.

Recently I had to add checks to the image upload for avatars which were doing the same thing

**Comment by lachlan-00** [comment]
I can reproduce this so i'm working on it as a bug now.

**Comment by lachlan-00** [comment]
It's something in the fileupload jquery from first glance.
https://github.com/blueimp/jQuery-File-Upload

The code in the upload pages isn't even getting to upload.class. just investigating.

**Comment by lachlan-00** [comment]
Script path is different. This fixes it i just need to test. the actual uploads now
![image](https://user-images.githubusercontent.com/1305249/67345127-46db3900-f57d-11e9-9f7b-25f6d4ddd433.png)

**Comment by lachlan-00** [comment]
edge branch has working uploads but i'm going to extend this a bit further to make a few more fixes before it goes into develop

There are a few things about the way files are inserted that i don't like and i have a few issues how uploads are handled in general.

**Comment by NatureMuncher** [comment]
Thank you so much for looking into this @lachlan-00! While you're in there, would it be reasonable to modify the "Add" function under Show Catalog to only update files which were modified in the past 24 hrs or so? Or maybe it's more appropriate to create a new item for the drop-down such as "Add last day's uploads" or something like that?
We have 20K songs in the database and it would be much more efficient to only look at recent files rather than all files.

**Comment by lachlan-00** [comment]
I know what you mean. I have 265k files. But I've actually already got a plan for this.

* In API 400001 I've added update_from_tags. This lets you update a single song/artist songs/album songs from their tags 

* I've also started working on a python library to that is compatible with this version of the API. https://github.com/lachlan-00/python3-ampache

* Once I'm finished with the API and library. I'll be porting all my personal python scripts into a new project where one of those files will search and update recently modified files only. https://github.com/lachlan-00/ampache-scripts
